### PR TITLE
feat(ui): add experiment flag store for experiment ai page

### PR DESCRIPTION
**Pull request by wwayne**
The store is used for controlling experimental AI features
The data is saved into localStorage

The PR for the UI part: https://github.com/TabbyML/tabby/pull/1724
屏幕截图 2024-03-26 113711

Close TAB-566

### CODE REVIEW

File: ee/tabby-ui/lib/stores/utils.ts [hunk 1]
```diff
@@ -7,3 +10,52 @@ export const getChatById = (
   if (!Array.isArray(chats) || !chatId) return undefined
   return chats.find(c => c.id === chatId)
 }
+
+// @reference:
+// https://github.com/pmndrs/zustand/blob/main/docs/integrations/persisting-store-data.md#hydration-and-asynchronous-storages
+export const createStoreWithHydrated = ({
```
Comment (liangfung) [inline ee/tabby-ui/lib/stores/utils.ts:16]: Use generics to make the return value typed

Comment (wwayne) [reply ee/tabby-ui/lib/stores/utils.ts]: Updated to `export const createStoreWithHydrated = <T extends Record<string, any>>({`

File: ee/tabby-ui/lib/stores/exp-flag-store.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import { createStoreWithHydrated } from './utils'
+
+interface InitialState {
```
Comment (liangfung) [inline ee/tabby-ui/lib/stores/exp-flag-store.ts:3]: Maybe it could be renamed to something like 'ExperimentalFlagState'?

Comment (wwayne) [reply ee/tabby-ui/lib/stores/exp-flag-store.ts:3]: yeah should rename

File: ee/tabby-ui/.eslintrc.json [hunk 3]
```diff
@@ -8,6 +8,7 @@
   ],
   "plugins": [
     "tailwindcss",
+    "@typescript-eslint",
```
Comment (wwayne) [inline ee/tabby-ui/.eslintrc.json:11]: Without `@typescript-eslint`, I got the following error when using Enum
```
Cannot read properties of undefined (reading 'loc')
Occurred while linting /home/wayne/work/tabby/ee/tabby-ui/lib/constants/index.ts:1
Rule: "unused-imports/no-unused-imports"
```

File: ee/tabby-ui/lib/constants/index.ts [hunk 4]
```diff
@@ -1,3 +1,7 @@
 export const PLACEHOLDER_EMAIL_FORM = 'name@yourcompany.com'
 
 export const DEFAULT_PAGE_SIZE = 20
+
+export enum EXPERIMENTAL_FALG {
+  'QUICK_ACTION_BAR' = 'quickActionBar'
+}
```
Comment (wsxiaoys) [inline ee/tabby-ui/lib/constants/index.ts:7]:
```suggestion

```

File: ee/tabby-ui/lib/stores/exp-flag-actions.ts
Comment (wsxiaoys) [inline ee/tabby-ui/lib/stores/exp-flag-actions.ts:1]: merge into single file as `tabby-ui/lib/experimental-flags.ts`

File: ee/tabby-ui/lib/stores/exp-flag-store.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import { EXPERIMENTAL_FALG } from '@/lib/constants'
+
+import { createStoreWithHydrated } from './utils'
+
+interface ExperimentalFlagInitialState {
+  [EXPERIMENTAL_FALG.QUICK_ACTION_BAR]: boolean
+}
+
```
Comment (wsxiaoys) [inline ee/tabby-ui/lib/stores/exp-flag-store.ts:8]: flags need to be accessible as global state, e.g
```
export EXP_enable_code_browser_quick_action_bar = ...
```

File: ee/tabby-ui/lib/stores/exp-flag-store.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+export const useExperimentalFlagStore =
+  createStoreWithHydrated<ExperimentalFlagInitialState>({
```
Comment (wsxiaoys) [inline ee/tabby-ui/lib/stores/exp-flag-store.ts:10]: it's fine to use zustand to create persistant store. But i generally feel it's more lightweight to use something like `useLocalStorage()` to wrap each individual flag as a pair of hook (plus global variable)

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

File: ee/tabby-ui/lib/stores/exp-flag-store.ts (hunk 5 above)
Comment (wsxiaoys) [reply ee/tabby-ui/lib/stores/exp-flag-store.ts:8]: Prototype api

```
const [EXP_enable_code_browser_quick_action_bar, useEnableCodeBrowserQuickActionBar] = defineExperimentalFlag(/* storage_key */ "enable_code_browser_quick_action_bar", /* default_value */ false, /* description */ "This flag is use for ...");

export {
  EXP_enable_code_browser_quick_action_bar,
  useEnableCodeBrowserQuickActionBar
}

```

File: ee/tabby-ui/lib/stores/exp-flag-actions.ts
Comment (wwayne) [reply ee/tabby-ui/lib/stores/exp-flag-actions.ts:1]: moved

File: ee/tabby-ui/lib/stores/exp-flag-store.ts (hunk 6 above)
Comment (wwayne) [reply ee/tabby-ui/lib/stores/exp-flag-store.ts]: I did a refactory, plz take a check

File: ee/tabby-ui/lib/constants/index.ts (hunk 4 above)
Comment (wwayne) [reply ee/tabby-ui/lib/constants/index.ts]: removed

File: ee/tabby-ui/lib/experimental-flags.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  }
+}
+
+const defineExperimentalFlagHook = (
+  storageKey: string,
+  flag: ExperimentalFlag
+) => {
+  return () => {
+    const [storageValue, setStorageValue] = useLocalStorage(
+      storageKey,
```
Comment (wsxiaoys) [inline ee/tabby-ui/lib/experimental-flags.ts:39]: add prefix `EXP_`

File: ee/tabby-ui/lib/experimental-flags.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
+import useLocalStorage from 'use-local-storage'
+
+type DefineExperimentalFlagResponse = [
+  ExperimentalFlag,
+  () => {
```
Comment (wsxiaoys) [inline ee/tabby-ui/lib/experimental-flags.ts:5]: return array of flag and toggle function

File: ee/tabby-ui/lib/experimental-flags.ts [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+  storageKey: string,
+  description: string,
+  defaultValue?: boolean
+): DefineExperimentalFlagResponse => {
+  const flagDefaultValue = defaultValue ?? false
+  const flag = new ExperimentalFlag(storageKey, description, flagDefaultValue)
+  const useFlagHook = defineExperimentalFlagHook(storageKey, flag)
+  return [flag, useFlagHook]
+}
+
```
Comment (wsxiaoys) [inline ee/tabby-ui/lib/experimental-flags.ts:65]:
```
const enableCodeBrowserQuickActionFlag = ....
export const EXP_enable_code_browser_quick_action_bar = enableCodeBrowserQuickActionFlag.defineGlobal();
export const useEnableCodeBrowserQuickActionBar = enableCodeBrowserQuickActionFlag.defineHook();
```

Comment (wwayne) [reply ee/tabby-ui/lib/experimental-flags.ts:63]:
```js
const quickActionBarFlag = new RuntimeFlagFactory(
  'enable_code_browser_quick_action_bar',
  'Show a quick action popup upon selecting code snippets',
  false
)

export const EXP_enable_code_browser_quick_action_bar = quickActionBarFlag.defineGlobalVarAccess()
export const useEnableCodeBrowserQuickActionBar = quickActionBarFlag.defineHookAccess()
```

File: ee/tabby-ui/lib/experimental-flags.ts (hunk 7 above)
Comment (wwayne) [reply ee/tabby-ui/lib/experimental-flags.ts]: Added in here
```
class RuntimeFlagFactory {
  private storageKey: string
  private description: string
  private defaultValue: boolean

  constructor(
    storageKey: string,
    description: string,
    defaultValue?: boolean
  ) {
    this.storageKey = `EXP_${storageKey}`
    this.description = description
    this.defaultValue = defaultValue ?? false
  }
```

File: ee/tabby-ui/lib/experimental-flags.ts (hunk 8 above)
Comment (wwayne) [reply ee/tabby-ui/lib/experimental-flags.ts]: updated, the return type is 
```js
[
  { value: boolean, description: string },
  () => void
]
```

File: ee/tabby-ui/lib/experimental-flags.ts [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+          value: storageValue,
+          description: this.description
+        },
+        toggleFlag
+      ]
+    }
+  }
+}
+
+const quickActionBarFlag = new RuntimeFlagFactory(
```
Comment (wsxiaoys) [inline ee/tabby-ui/lib/experimental-flags.ts:57]:
```suggestion
const enableCodeBrowserQuickActionBarFactory = new RuntimeFlagFactory(
```

File: ee/tabby-ui/lib/experimental-flags.ts [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+  'enable_code_browser_quick_action_bar',
+  'Show a quick action popup upon selecting code snippets',
+  false
+)
+
+export const EXP_enable_code_browser_quick_action_bar =
+  quickActionBarFlag.defineGlobalVarAccess()
```
Comment (wsxiaoys) [inline ee/tabby-ui/lib/experimental-flags.ts:64]:
```suggestion
  quickActionBarFlag.defineGlobalVar()
```

File: ee/tabby-ui/lib/experimental-flags.ts [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+export const useEnableCodeBrowserQuickActionBar =
+  quickActionBarFlag.defineHookAccess()
```
Comment (wsxiaoys) [inline ee/tabby-ui/lib/experimental-flags.ts:66]:
```suggestion
  quickActionBarFlag.defineHook()
```

File: ee/tabby-ui/lib/stores/utils.ts [hunk 13]
```diff
@@ -7,3 +10,58 @@ export const getChatById = (
   if (!Array.isArray(chats) || !chatId) return undefined
   return chats.find(c => c.id === chatId)
 }
+
+// @reference:
+// https://github.com/pmndrs/zustand/blob/main/docs/integrations/persisting-store-data.md#hydration-and-asynchronous-storages
+export const createStoreWithHydrated = <T extends Record<string, any>>({
```
Comment (wsxiaoys) [inline ee/tabby-ui/lib/stores/utils.ts:16]: this can be removed?

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

File: ee/tabby-ui/lib/experimental-flags.ts [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+      if (storageValue) {
+        return storageValue === 'true'
+      }
+    }
+
+    return this.defaultValue
+  }
+}
+
+class RuntimeFlagFactory {
```
Comment (wsxiaoys) [inline ee/tabby-ui/lib/experimental-flags.ts:22]:
```suggestion
class ExperimentFlagFactory {
```

File: ee/tabby-ui/lib/experimental-flags.ts [hunk 15]
```diff
@@ -0,0 +1,66 @@
+import useLocalStorage from 'use-local-storage'
+
+class FeatureFlag {
```
Comment (wsxiaoys) [inline ee/tabby-ui/lib/experimental-flags.ts:3]:
```suggestion
class ExperimentFlag {
```

File: ee/tabby-ui/lib/experimental-flags.ts
Comment (wsxiaoys) [inline ee/tabby-ui/lib/experimental-flags.ts:1]: filename change to `experiment-flags.ts`

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

File: ee/tabby-ui/lib/stores/utils.ts (hunk 13 above)
Comment (wwayne) [reply ee/tabby-ui/lib/stores/utils.ts]: removed

**Comment by wwayne** [comment]
names updated

File: ee/tabby-ui/.eslintrc.json (hunk 3 above)
Comment (wsxiaoys) [inline ee/tabby-ui/.eslintrc.json:11]: is this still required? consider remove if it's not longer necessary

Comment (wwayne) [reply ee/tabby-ui/.eslintrc.json:11]: Without this plugin, the `unused-imports/no-unused-imports` will stop working when we using `Enum`

**Pull request merged**